### PR TITLE
Jacobian Support for Basic Network

### PR DESCRIPTION
**Pull request by jbarberia**
#762

I modified the commented out version of the function that calculates the Jacobian matrix for basic networks. Also add a function to compute the sub-matrices dP/dva and dQ/dvm.

I move the from scratch AC power flow solver to `test/common.jl` because the Jacobian matrix is tested with it.

### CODE REVIEW

**Comment by ccoffrin** [comment]
After another quick review, nothing stands out to me as notably problematic.  I might rename some of the local variables to be more inline with the code conventions but that is a minor point as long as the code is correct.

Some requests regarding the unit tests,
- The test for `calc_basic_jacobian_matrix` should include a case where a network is loaded in, all buses are changed to p/q and then `calc_basic_jacobian_matrix` is called.  I suggest using `case5.m` as the base network. 
- All of the `calc_basic_jacobian_matrix` tests should include some metric that captures the full matrix, so that we can detect changes that do not occur in specific cells. `LinearAlgebra.det` seems to work well for the admittance matrix tests.
- We need to cross validate the results with what comes out of Matpower.  If you are not able to do this I can help.

**Comment by jbarberia** [comment]
The suggestions made seem fine to me.

* I like the idea of testing in a case with all p/q buses.
* I'm going to add the determinant in the tests.
* How would the correlation with MATPOWER be?
[Here](https://github.com/lanl-ansi/PowerModels.jl/pull/777/commits/b5611b9b960f28e7fd90a422d3f45f290e824976#diff-014eb8a32b127fcf3633944b967ee8a2f0f26c56d53cbe4cb8e9db76d31f966eR193) I simply took the values of the Jacobian made in MATPOWER.

Tomorrow I can add a commit whit this.

**Comment by ccoffrin** [comment]
As long as you did a side by side check with the output of Matpower's output, then we are good.

**Comment by jbarberia** [comment]
@ccoffrin, I had a problem with the unit test for the case 5:

The workflow that I do the side by side comparison with Matpower was:
```julia
using PowerModels
data = make_basic_network(PowerModels.parse_file("../test/data/matpower/case5.m"))
for (i, bus) in data["bus"] # All buses PQ
    bus["bus_type"] = 1
end
export_to_matpower("case5basic.m", data)
J = calc_basic_jacobian_matrix(data)
```
and then make the Jacobian with Matpower:
```matlab
J = full(makeJac((case5basic), 1)) % The 1 is for the full jacobian
```

It was strange to notice that the determinant of the Jacobian in Julia and in Matlab gave very different values ​​for case 5. Even if the matrices were the same!

I checked it using the `MAT` module to load the Jacobian matrix from Matpower to Julia
```julia
using MAT
J_matpower = matread("J.mat")["J"]
for i in 1:10
    for j in 1:10
        @test isapprox(J[i, j], J_matpower[i, j])
    end
end 
```

File: test/data-basic.jl [hunk 1]
```diff
@@ -189,53 +189,74 @@ end
...
+            for (i, bus) in data["bus"] # All buses PQ
+                bus["bus_type"] = 1
+            end
+            J = PowerModels.calc_basic_jacobian_matrix(data);
+
+            num_bus = length(data["bus"])
+            
+            @test size(J)[2] == num_bus * 2
+            @test size(J)[1] == num_bus * 2
+            @test isapprox(LinearAlgebra.det(J), 1524.5588; atol=1e-4)
```
Comment (jbarberia) [inline test/data-basic.jl:204]: The `LinearAlgebra.det` is not a good full matrix metric.
``` julia
J = calc_basic_jacobian_matrix(data) # J :: SparseMatrixCSC
J |> det # 1524.55...
J |> Matrix |> det # -1590.68...
U, V = svd(J |> Matrix); reduce(*, V) # 3967.00...
```

What do you think about something simple like:
``` julia
J = calc_basic_jacobian_matrix(data) # J :: SparseMatrixCSC
metric(arr) = reduce(*, [sum(row) for row in eachcol(arr)])
J |> metric # -132063.9683811387
J |> Matrix |> metric # -132063.9683811387
```

Comment (ccoffrin) [reply test/data-basic.jl:204]: I an fine with the sum metric instead.

**Comment by jbarberia** [comment]
@ccoffrin, Could you review the code to be able to merge the PR?

**Comment by jbarberia** [comment]
@ccoffrin, Should I close this PR?

**Comment by ccoffrin** [comment]
No way!  Sorry for the delay, I do plan to merge this hopefully before the next patch release.

Were you in a hurry to get it to be part of the PowerModels release?

**Comment by jbarberia** [comment]
Don't worry. I ain't hurry with this PR. I only wanted to know if it was useful to leave it open or not.

**Comment by ccoffrin** [comment]
Hi @jbarberia concerning the fast decoupled solver, I like the implementation that you propose because it is algorithmically the same as the fully coupled case.  However, as I recall, the traditional fast decoupled load flow has a structure of update, Theta, recompute, update V, recompute, update Theta, ...; rather than update Theta/V, recompute, update Theta/V, ...

Do you know if there is some president/reference for the approach you have taken here?

**Comment by ccoffrin** [comment]
@jbarberia, can you add yourself to the README.md file as a contributor of the basic network Jacobean feature?

**Comment by jbarberia** [comment]
@ccoffrin, for sure!

**Pull request merged**